zkevm-circuits/src/evm_circuit/step.rs

```diff
@@ -121,6 +125,154 @@ impl Display for ExecutionState {
+                    OpcodeId::CALLER => ExecutionState::CALLER,
+                    OpcodeId::CALLVALUE => ExecutionState::CALLVALUE,
+                    OpcodeId::EXTCODEHASH => ExecutionState::EXTCODEHASH,
+                    OpcodeId::EXTCODESIZE => ExecutionState::EXTCODESIZE,
+                    OpcodeId::BLOCKHASH => ExecutionState::BLOCKHASH,
+                    OpcodeId::TIMESTAMP | OpcodeId::NUMBER | OpcodeId::GASLIMIT => {
+                        ExecutionState::BLOCKCTXU64
+                    }
+                    OpcodeId::COINBASE => ExecutionState::BLOCKCTXU160,
+                    OpcodeId::DIFFICULTY | OpcodeId::BASEFEE => ExecutionState::BLOCKCTXU256,
+                    OpcodeId::GAS => ExecutionState::GAS,
+                    OpcodeId::SAR => ExecutionState::SAR,
+                    OpcodeId::SELFBALANCE => ExecutionState::SELFBALANCE,
+                    OpcodeId::SHA3 => ExecutionState::SHA3,
+                    OpcodeId::SHL | OpcodeId::SHR => ExecutionState::SHL_SHR,
+                    OpcodeId::SLOAD => ExecutionState::SLOAD,
+                    OpcodeId::SSTORE => ExecutionState::SSTORE,
+                    OpcodeId::CALLDATASIZE => ExecutionState::CALLDATASIZE,
+                    OpcodeId::CALLDATACOPY => ExecutionState::CALLDATACOPY,
+                    OpcodeId::CHAINID => ExecutionState::CHAINID,
+                    OpcodeId::ISZERO => ExecutionState::ISZERO,
+                    OpcodeId::CALL
+                    | OpcodeId::CALLCODE
+                    | OpcodeId::DELEGATECALL
+                    | OpcodeId::STATICCALL => ExecutionState::CALL_OP,
+                    OpcodeId::ORIGIN => ExecutionState::ORIGIN,
+                    OpcodeId::CODECOPY => ExecutionState::CODECOPY,
+                    OpcodeId::CALLDATALOAD => ExecutionState::CALLDATALOAD,
+                    OpcodeId::CODESIZE => ExecutionState::CODESIZE,
+                    OpcodeId::EXTCODECOPY => ExecutionState::EXTCODECOPY,
+                    OpcodeId::RETURN | OpcodeId::REVERT => ExecutionState::RETURN_REVERT,
+                    OpcodeId::RETURNDATASIZE => ExecutionState::RETURNDATASIZE,
+                    OpcodeId::RETURNDATACOPY => ExecutionState::RETURNDATACOPY,
+                    // dummy ops
+                    OpcodeId::CREATE => dummy!(ExecutionState::CREATE),
+                    OpcodeId::CREATE2 => dummy!(ExecutionState::CREATE2),
+                    OpcodeId::SELFDESTRUCT => dummy!(ExecutionState::SELFDESTRUCT),
+                    _ => unimplemented!("unimplemented opcode {:?}", op),
+                }
+            }
+            ExecState::BeginTx => ExecutionState::BeginTx,
+            ExecState::EndTx => ExecutionState::EndTx,
+            ExecState::EndBlock => ExecutionState::EndBlock,
+        }
+    }
+}
+
+pub trait HasExecutionState {
+    fn execution_state(&self) -> ExecutionState;
+}
+
+impl HasExecutionState for ExecStep {
+    fn execution_state(&self) -> ExecutionState {
```

Comment:
I see that the trait `HasExecutionState` is only implemented for `ExecStep`.  Did you implement `execution_state` via a trait in order to keep this implementation in the zkevm-circuits crate?  Did you consider implementing it directly in the `bus-mapping` crate so that no trait is needed?

Reply:
The problem is that `ExecutionState` is only in zkevm-circuit crate, so I can not implement the `execution_state` method for the `ExecStep` in the `bus-mapping` crate.

Reply:
I see, thanks for the clarification!